src/action/list_directory.rs

```diff
@@ -0,0 +1,730 @@
+    let mut linux_flags: c_long = 0;
+    let linux_flags_ptr: *mut c_long = &mut linux_flags;
+    unsafe {
+        match ioctls::fs_ioc_getflags(file.as_raw_fd(), linux_flags_ptr) {
+            0 => Some(linux_flags as u32),
+            _ => None,
+        }
+    }
+}
+
+impl super::Request for Request {
+
+    type Proto = ListDirRequest;
+
+    fn from_proto(proto: Self::Proto) -> Result<Request, session::ParseError> {
+        let missing = session::MissingFieldError::new;
+        let pathspec = proto.pathspec.ok_or(missing("path spec"))?;
+        let path_type = pathspec.pathtype
+            .ok_or(missing("path type"))?;
+        if path_type != PathType::Os as i32 {
+            return Err(ParseError::unsupported_value(String::from("path type"),
+                                                     path_type.to_string()));
+        }
+        let path_option = pathspec.path_options
+            .unwrap_or(Options::CaseLiteral as i32);
+        if path_option != Options::CaseLiteral as i32 {
+            return Err(ParseError::unsupported_value(
+                String::from("path option"),
+                path_option.to_string()));
+        };
+        Ok(Request {
+            path: get_path(&pathspec.path),
+        })
+    }
+}
+
+/// Converts idiomatic `SystemTime` to `u64` for the protocol buffer.
+fn get_time_since_unix_epoch(sys_time: &Option<SystemTime>) -> Option<u64> {
+    match sys_time {
+        Some(time) => match micros(time.clone()) {
+            Ok(time) => Some(time),
+            Err(error) => {
+                warn!("failed to convert time: {}", error);
+                None
+            }
+        }
+        None => None,
+    }
+}
+
+impl super::Response for Response {
+
+    const RDF_NAME: Option<&'static str> = Some("StatEntry");
+
+    type Proto = StatEntry;
+
+    fn into_proto(self) -> Self::Proto {
+        StatEntry {
+            st_mode: self.mode,
+            st_ino: self.ino,
+            st_dev: self.dev,
+            st_nlink: self.nlink,
+            st_uid: self.uid,
+            st_gid: self.gid,
+            st_size: Some(self.size),
+            st_atime: get_time_since_unix_epoch(&self.atime),
+            st_mtime: get_time_since_unix_epoch(&self.mtime),
+            st_ctime: get_time_since_unix_epoch(&self.ctime),
+            st_blocks: self.blocks,
+            st_blksize: self.blksize,
+            st_rdev: self.rdev,
+            st_flags_osx: None,
+            st_flags_linux: self.flags_linux,
+            symlink:
+            self.symlink
+                .map(|symlink| symlink.to_string_lossy().to_string()),
+            registry_type: None,
+            resident: None,
+            pathspec: Some(rrg_proto::PathSpec {
+                // Represents `CaseLiteral` path option (other options are not
+                // supported).
+                path_options: Some(Options::CaseLiteral as i32),
+                // Represents OS path type (other types are not supported).
```

Comment:
I cannot reply to the previous thread (I can make only mark it as resolved), so I will just reply here: good point about "OS" also meaning an enum-specific value. In that case, I am not sure what is better after all.